# Data Cleaning and Transformation
## Notebook 1

Group project by:
    - Tinuke Durotolu
    - Sara Ruini
    - Susan Yousefi


<b> Data source: </b> https://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008# 

<b> Reference: </b>
Beata Strack, Jonathan P. DeShazo, Chris Gennings, Juan L. Olmo, Sebastian Ventura, Krzysztof J. Cios, and John N. Clore, “Impact of HbA1c Measurement on Hospital Readmission Rates: Analysis of 70,000 Clinical Database Patient Records,” BioMed Research International, vol. 2014, Article ID 781670, 11 pages, 2014. 

In [1]:
#libraries needed
import pandas as pd
import numpy as np
import os

## Dataset challenges
This dataset has been pre-cleaned. However, there's some modifications that we need to perform.

### About the medicines
# ADD SOURCE?!?!??!??!??!?!!? ->

    - Metformin is a biguanide that lowers blood sugar levels.
    - Repaglinide and Nateglinide belongs to metglinides drug class and also helps lower blood glucose levels.
    - Tolbutamide, glimepiride and glipizide are sulfonylureas that help stimulate pancreas to produce insulin.
    - Pioglitazine is a thiazolidinedione and increased the sensitivity of body cells to insulin.
    - The other groups include gliptins, insulins and the other medications we have grouped as others.
    
In the column of each medicine (e.g. 'metmorfin'), "Up" means that the medicine got incresed, "Down" means that it got decreased, "Steady" means that the dosage has not changed and "No" means it was not prescribed.

### This dataset contains...

- the main diabetic data file ("diabetic_data_csv")
- the ID mapping to "admission_type_id" ("IDs_mapping.csv").

For convenience, we are going to main dataset 'df' and the ids 'ids'.

In [2]:
df = pd.read_csv("diabetic_data.csv", encoding="Latin-1")
ids = pd.read_csv("IDs_mapping.csv", encoding="Latin-1")

## Visualising the datasets

First of all, we are going to visualise the datatype of each column in the dataset and see if there is any missing values.

Our aim is to remove unnecessary values and columns and remove any information that we consider redundant.

## IDs_mapping dataframe analysis('ids')

In [3]:
ids

admission_type_id  \
0                          1   
1                          2   
2                          3   
3                          4   
4                          5   
5                          6   
6                          7   
7                          8   
8                        NaN   
9   discharge_disposition_id   
10                         1   
11                         2   
12                         3   
13                         4   
14                         5   
15                         6   
16                         7   
17                         8   
18                         9   
19                        10   
20                        11   
21                        12   
22                        13   
23                        14   
24                        15   
25                        16   
26                        17   
27                        18   
28                        19   
29                        20   
..                       ...   
37                        27   
38                        28   
39                        29   
40                       NaN   
41       admission_source_id   
42                         1   
43                         2   
44                         3   
45                         4   
46                         5   
47                         6   
48                         7   
49                         8   
50                         9   
51                        10   
52                        11   
53                        12   
54                        13   
55                        14   
56                        15   
57                        17   
58                        18   
59                        19   
60                        20   
61                        21   
62                        22   
63                        23   
64                        24   
65                        25   
66                        26   

                                          description  
0                                           Emergency  
1                                              Urgent  
2                                            Elective  
3                                             Newborn  
4                                       Not Available  
5                                                 NaN  
6                                       Trauma Center  
7                                          Not Mapped  
8                                                 NaN  
9                                         description  
10                                 Discharged to home  
11  Discharged/transferred to another short term h...  
12                      Discharged/transferred to SNF  
13                      Discharged/transferred to ICF  
14  Discharged/transferred to another type of inpa...  
15  Discharged/transferred to home with home healt...  
16                                           Left AMA  
17  Discharged/transferred to home under care of H...  
18          Admitted as an inpatient to this hospital  
19  Neonate discharged to another hospital for neo...  
20                                            Expired  
21  Still patient or expected to return for outpat...  
22                                     Hospice / home  
23                         Hospice / medical facility  
24  Discharged/transferred within this institution...  
25  Discharged/transferred/referred another instit...  
26  Discharged/transferred/referred to this instit...  
27                                                NaN  
28           Expired at home. Medicaid only, hospice.  
29  Expired in a medical facility. Medicaid only, ...  
..                                                ...  
37  Discharged/transferred to a federal health car...  
38  Discharged/transferred/referred to a psychiatr...  
39  Discharged/transferred to a Critical Access Ho...  
40                                                NaN  
41        

We can see that there's three ids being mapped: admission_type_id, discharge_disposition_id, admission_source_id.

We will only keep the first two ids and save them into separate dataframes.

## Removing unnecessary values

In [4]:
#removing values for 'admission_source_id' as they are noted needed
ids = ids.drop(ids.index[list(np.arange(41,67))])
ids_nulls = ids[ids.isnull().any(axis=1)]

In [5]:
ids_nulls

admission_type_id description
5                  6         NaN
8                NaN         NaN
27                18         NaN
40               NaN         NaN

Later, we will delete values 6 and NaN from 'admission_type_id', and 18 and 40 from 'discharge_disposition_id'

In [6]:
ids

admission_type_id  \
0                          1   
1                          2   
2                          3   
3                          4   
4                          5   
5                          6   
6                          7   
7                          8   
8                        NaN   
9   discharge_disposition_id   
10                         1   
11                         2   
12                         3   
13                         4   
14                         5   
15                         6   
16                         7   
17                         8   
18                         9   
19                        10   
20                        11   
21                        12   
22                        13   
23                        14   
24                        15   
25                        16   
26                        17   
27                        18   
28                        19   
29                        20   
30                        21   
31                        22   
32                        23   
33                        24   
34                        25   
35                        26   
36                        30   
37                        27   
38                        28   
39                        29   
40                       NaN   

                                          description  
0                                           Emergency  
1                                              Urgent  
2                                            Elective  
3                                             Newborn  
4                                       Not Available  
5                                                 NaN  
6                                       Trauma Center  
7                                          Not Mapped  
8                                                 NaN  
9                                         description  
10                                 Discharged to home  
11  Discharged/transferred to another short term h...  
12                      Discharged/transferred to SNF  
13                      Discharged/transferred to ICF  
14  Discharged/transferred to another type of inpa...  
15  Discharged/transferred to home with home healt...  
16                                           Left AMA  
17  Discharged/transferred to home under care of H...  
18          Admitted as an inpatient to this hospital  
19  Neonate discharged to another hospital for neo...  
20                                            Expired  
21  Still patient or expected to return for outpat...  
22                                     Hospice / home  
23                         Hospice / medical facility  
24  Discharged/transferred within this institution...  
25  Discharged/transferred/referred another instit...  
26  Discharged/transferred/referred to this instit...  
27                                                NaN  
28           Expired at home. Medicaid only, hospice.  
29  Expired in a medical facility. Medicaid only, ...  
30    Expired, place unknown. Medicaid only, hospice.  
31  Discharged/transferred to another rehab fac in...  
32  Discharged/transferred to a long term care hos...  
33  Discharged/transferred to a nursing facility c...  
34                                         Not Mapped  
35                                    Unknown/Invalid  
36  Discharged/transferred to another Type of Heal...  
37  Discharged/transferred to a federal health car...  
38  Discharged/transferred/referred to a psychiatr...  
39  Discharged/transferred to a Critical Access Ho...  
40                                                NaN

In [7]:
discharge_disposition_id_df = ids.loc[10:].copy()
#separating values for discharge_disposition_id
admission_type_id_df = ids.loc[:8].copy()
admission_type_id_df                                

admission_type_id    description
0                 1      Emergency
1                 2         Urgent
2                 3       Elective
3                 4        Newborn
4                 5  Not Available
5                 6            NaN
6                 7  Trauma Center
7                 8     Not Mapped
8               NaN            NaN

In [8]:
admission_type_id_df = admission_type_id_df.dropna()
admission_type_id_df

admission_type_id    description
0                 1      Emergency
1                 2         Urgent
2                 3       Elective
3                 4        Newborn
4                 5  Not Available
6                 7  Trauma Center
7                 8     Not Mapped

In [9]:
discharge_disposition_id_df = discharge_disposition_id_df.dropna()
discharge_disposition_id_df = discharge_disposition_id_df.rename(columns={"admission_type_id": "discharge_disposition_id"})
discharge_disposition_id_df

discharge_disposition_id                                        description
10                        1                                 Discharged to home
11                        2  Discharged/transferred to another short term h...
12                        3                      Discharged/transferred to SNF
13                        4                      Discharged/transferred to ICF
14                        5  Discharged/transferred to another type of inpa...
15                        6  Discharged/transferred to home with home healt...
16                        7                                           Left AMA
17                        8  Discharged/transferred to home under care of H...
18                        9          Admitted as an inpatient to this hospital
19                       10  Neonate discharged to another hospital for neo...
20                       11                                            Expired
21                       12  Still patient or expected to return for outpat...
22                       13                                     Hospice / home
23                       14                         Hospice / medical facility
24                       15  Discharged/transferred within this institution...
25                       16  Discharged/transferred/referred another instit...
26                       17  Discharged/transferred/referred to this instit...
28                       19           Expired at home. Medicaid only, hospice.
29                       20  Expired in a medical facility. Medicaid only, ...
30                       21    Expired, place unknown. Medicaid only, hospice.
31                       22  Discharged/transferred to another rehab fac in...
32                       23  Discharged/transferred to a long term care hos...
33                       24  Discharged/transferred to a nursing facility c...
34                       25                                         Not Mapped
35                       26                                    Unknown/Invalid
36                       30  Discharged/transferred to another Type of Heal...
37                       27  Discharged/transferred to a federal health car...
38                       28  Discharged/transferred/referred to a psychiatr...
39                       29  Discharged/transferred to a Critical Access Ho...

## diabetic_data dataframe analysis('df')
We will visualise the info of the dataset (presence of null-values) for each column, show the head and tail of the dataset and show the percentage of unique value_counts. 

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
encounter_id                101766 non-null int64
patient_nbr                 101766 non-null int64
race                        101766 non-null object
gender                      101766 non-null object
age                         101766 non-null object
weight                      101766 non-null object
admission_type_id           101766 non-null int64
discharge_disposition_id    101766 non-null int64
admission_source_id         101766 non-null int64
time_in_hospital            101766 non-null int64
payer_code                  101766 non-null object
medical_specialty           101766 non-null object
num_lab_procedures          101766 non-null int64
num_procedures              101766 non-null int64
num_medications             101766 non-null int64
number_outpatient           101766 non-null int64
number_emergency            101766 non-null int64
number_inpatient            10176

There are a total of 101766 data entries in this dataset.

Before analysing this dataframe further, we will drop unnecessary columns. Specifically, we will delete medicines prescribed that are not used in the UK and some variables that are not needed for our model.

In [11]:
to_drop = ['weight','medical_specialty','max_glu_serum', 'admission_source_id', 'encounter_id','patient_nbr', 'diag_1', 'diag_2', 'diag_3', 'payer_code', 'A1Cresult','glyburide-metformin',  'chlorpropamide',  'tolazamide', 'glipizide-metformin', 'glimepiride-pioglitazone','metformin-rosiglitazone','metformin-pioglitazone', 'nateglinide', 'examide', 'rosiglitazone','miglitol', 'acetohexamide', 'glipizide', 'glyburide', 'troglitazone']
df.drop(to_drop, inplace=True, axis=1)

In [12]:
df.head(5)

race  gender      age  admission_type_id  \
0        Caucasian  Female   [0-10)                  6   
1        Caucasian  Female  [10-20)                  1   
2  AfricanAmerican  Female  [20-30)                  1   
3        Caucasian    Male  [30-40)                  1   
4        Caucasian    Male  [40-50)                  1   

   discharge_disposition_id  time_in_hospital  num_lab_procedures  \
0                        25                 1                  41   
1                         1                 3                  59   
2                         1                 2                  11   
3                         1                 2                  44   
4                         1                 1                  51   

   num_procedures  num_medications  number_outpatient  ...  repaglinide  \
0               0                1                  0  ...           No   
1               0               18                  0  ...           No   
2               5               13                  2  ...           No   
3               1               16                  0  ...           No   
4               0                8                  0  ...           No   

   glimepiride  tolbutamide pioglitazone acarbose citoglipton insulin change  \
0           No           No           No       No          No      No     No   
1           No           No           No       No          No      Up     Ch   
2           No           No           No       No          No      No     No   
3           No           No           No       No          No      Up     Ch   
4           No           No           No       No          No  Steady     Ch   

  diabetesMed readmitted  
0          No         NO  
1         Yes        >30  
2         Yes         NO  
3         Yes         NO  
4         Yes         NO  

[5 rows x 24 columns]

In [13]:
df.tail(5)

race  gender      age  admission_type_id  \
101761  AfricanAmerican    Male  [70-80)                  1   
101762  AfricanAmerican  Female  [80-90)                  1   
101763        Caucasian    Male  [70-80)                  1   
101764        Caucasian  Female  [80-90)                  2   
101765        Caucasian    Male  [70-80)                  1   

        discharge_disposition_id  time_in_hospital  num_lab_procedures  \
101761                         3                 3                  51   
101762                         4                 5                  33   
101763                         1                 1                  53   
101764                         3                10                  45   
101765                         1                 6                  13   

        num_procedures  num_medications  number_outpatient  ...  repaglinide  \
101761               0               16                  0  ...           No   
101762               3               18                  0  ...           No   
101763               0                9                  1  ...           No   
101764               2               21                  0  ...           No   
101765               3                3                  0  ...           No   

        glimepiride  tolbutamide pioglitazone acarbose citoglipton insulin  \
101761           No           No           No       No          No    Down   
101762           No           No           No       No          No  Steady   
101763           No           No           No       No          No    Down   
101764           No           No       Steady       No          No      Up   
101765           No           No           No       No          No      No   

       change diabetesMed readmitted  
101761     Ch         Yes        >30  
101762     No         Yes         NO  
101763     Ch         Yes         NO  
101764     Ch         Yes         NO  
101765     No          No         NO  

[5 rows x 24 columns]

In [14]:
df = df.dropna(axis=0, how="any")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 101766 entries, 0 to 101765
Data columns (total 24 columns):
race                        101766 non-null object
gender                      101766 non-null object
age                         101766 non-null object
admission_type_id           101766 non-null int64
discharge_disposition_id    101766 non-null int64
time_in_hospital            101766 non-null int64
num_lab_procedures          101766 non-null int64
num_procedures              101766 non-null int64
num_medications             101766 non-null int64
number_outpatient           101766 non-null int64
number_emergency            101766 non-null int64
number_inpatient            101766 non-null int64
number_diagnoses            101766 non-null int64
metformin                   101766 non-null object
repaglinide                 101766 non-null object
glimepiride                 101766 non-null object
tolbutamide                 101766 non-null object
pioglitazone                1017

In [15]:
df.describe(include='all')

race  gender      age  admission_type_id  \
count      101766  101766   101766      101766.000000   
unique          6       3       10                NaN   
top     Caucasian  Female  [70-80)                NaN   
freq        76099   54708    26068                NaN   
mean          NaN     NaN      NaN           2.024006   
std           NaN     NaN      NaN           1.445403   
min           NaN     NaN      NaN           1.000000   
25%           NaN     NaN      NaN           1.000000   
50%           NaN     NaN      NaN           1.000000   
75%           NaN     NaN      NaN           3.000000   
max           NaN     NaN      NaN           8.000000   

        discharge_disposition_id  time_in_hospital  num_lab_procedures  \
count              101766.000000     101766.000000       101766.000000   
unique                       NaN               NaN                 NaN   
top                          NaN               NaN                 NaN   
freq                         NaN               NaN                 NaN   
mean                    3.715642          4.395987           43.095641   
std                     5.280166          2.985108           19.674362   
min                     1.000000          1.000000            1.000000   
25%                     1.000000          2.000000           31.000000   
50%                     1.000000          4.000000           44.000000   
75%                     4.000000          6.000000           57.000000   
max                    28.000000         14.000000          132.000000   

        num_procedures  num_medications  number_outpatient  ...  repaglinide  \
count    101766.000000    101766.000000      101766.000000  ...       101766   
unique             NaN              NaN                NaN  ...            4   
top                NaN              NaN                NaN  ...           No   
freq               NaN              NaN                NaN  ...       100227   
mean          1.339730        16.021844           0.369357  ...          NaN   
std           1.705807         8.127566           1.267265  ...          NaN   
min           0.000000         1.000000           0.000000  ...          NaN   
25%           0.000000        10.000000           0.000000  ...          NaN   
50%           1.000000        15.000000           0.000000  ...          NaN   
75%           2.000000        20.000000           0.000000  ...          NaN   
max           6.000000        81.000000          42.000000  ...          NaN   

        glimepiride  tolbutamide pioglitazone acarbose citoglipton insulin  \
count        101766       101766       101766   101766      101766  101766   
unique            4            2            4        4           1       4   
top              No           No           No       No          No      No   
freq          96575       101743        94438   101458      101766   47383   
mean            NaN          NaN          NaN      NaN         NaN     NaN   
std             NaN          NaN          NaN      NaN         NaN     NaN   
min             NaN          NaN          NaN      NaN         NaN     NaN   
25%             NaN          NaN          NaN      NaN         NaN     NaN   
50%             NaN          NaN          NaN      NaN         NaN     NaN   
75%             NaN          NaN          NaN      NaN         NaN     NaN   
max             NaN          NaN          NaN      NaN         NaN     NaN   

        change diabetesMed readmitted  
count   101766      101766     101766  
unique       2           2          3  
top         No         Yes         NO  
freq     54755       78363      54864  
mean       NaN         NaN        NaN  
std        NaN         NaN        NaN  
min        NaN         NaN        NaN  
25%        NaN         NaN        NaN  
50%        NaN         NaN        NaN  
75%        NaN         NaN        NaN  
max        NaN         NaN        NaN  

[11 rows x 24 columns]

In [16]:
def value_counter(dataset):
    for index in df.columns:
        print("---- Index: " + index + "----")
        print(df[index].value_counts(normalize=True, sort=True))
value_counter(df)

---- Index: race----
Caucasian          0.747784
AfricanAmerican    0.188766
?                  0.022336
Hispanic           0.020017
Other              0.014799
Asian              0.006299
Name: race, dtype: float64
---- Index: gender----
Female             0.537586
Male               0.462384
Unknown/Invalid    0.000029
Name: gender, dtype: float64
---- Index: age----
[70-80)     0.256156
[60-70)     0.220928
[50-60)     0.169565
[80-90)     0.168986
[40-50)     0.095169
[30-40)     0.037095
[90-100)    0.027445
[20-30)     0.016282
[10-20)     0.006790
[0-10)      0.001582
Name: age, dtype: float64
---- Index: admission_type_id----
1    0.530531
3    0.185416
2    0.181593
6    0.051992
5    0.047020
8    0.003144
7    0.000206
4    0.000098
Name: admission_type_id, dtype: float64
---- Index: discharge_disposition_id----
1     0.591887
3     0.137118
6     0.126781
18    0.036269
2     0.020911
22    0.019584
11    0.016135
5     0.011635
25    0.009718
4     0.008009
7     0.006122


Some columns have a lot of unique numeric values which could be put in ranges. Let's display them to make each category more readable.  These columns are: 'num_lab_procedures', 'num_medications', 

Later, we will also group them by category.

In [17]:
cols_to_range = ['num_lab_procedures', 'num_medications']
for col in cols_to_range:
    unique_val_col = df[col].unique().shape
    print("Column: '" + col + "'. Number of unique values: " + str(unique_val_col) + ".")
    print(df[col].value_counts(bins=10, sort=True))
    print("-------")

Column: 'num_lab_procedures'. Number of unique values: (118,).
(40.3, 53.4]      27814
(27.2, 40.3]      21826
(53.4, 66.5]      20558
(0.868, 14.1]     10318
(14.1, 27.2]      10141
(66.5, 79.6]       8876
(79.6, 92.7]       1915
(92.7, 105.8]       290
(105.8, 118.9]       23
(118.9, 132.0]        5
Name: num_lab_procedures, dtype: int64
-------
Column: 'num_medications'. Number of unique values: (75,).
(9.0, 17.0]     45079
(17.0, 25.0]    24813
(0.919, 9.0]    20515
(25.0, 33.0]     7961
(33.0, 41.0]     2183
(41.0, 49.0]      730
(49.0, 57.0]      320
(57.0, 65.0]      131
(65.0, 73.0]       29
(73.0, 81.0]        5
Name: num_medications, dtype: int64
-------


## List of modifications that we need to perform

At first glance, we have agreed that we need to perform the following changes to the main dataset:

#### 1. Map the description with the value from "admission_type_id" and "discharge_disposition_ids"
#### 2. Remove missing/unclear values.
These include:
    - from 'admission_type_id': 5, 8, 27, 40, 57
    - from 'gender': 'Unknown/Invalid'
    - from 'race': '?'

#### 3. Renaming columns
The goal is to:
    - Clarify what the column in the data is;
    - Conform to python standards (e.g. using snakecase)

## Data Modification

### Step 1. Map the description with the value from "admission_type_id" and "discharge_disposition_ids"

In [18]:
print(np.dtype(df['admission_type_id']))
print(np.dtype(admission_type_id_df['admission_type_id']))
print(np.dtype(admission_type_id_df['description']))

int64
object
object


In [19]:
print(type(df['admission_type_id']))
print(type(admission_type_id_df['description']))
print(type(df['admission_type_id']))

<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>


In [20]:
admission_type_id_df['admission_type_id'].astype(str)
df['admission_type_id'].astype(int)

0         6
1         1
2         1
3         1
4         1
5         2
6         3
7         1
8         2
9         3
10        1
11        2
12        1
13        1
14        3
15        1
16        1
17        1
18        1
19        3
20        2
21        2
22        2
23        2
24        3
25        1
26        1
27        1
28        2
29        1
         ..
101736    1
101737    3
101738    1
101739    1
101740    3
101741    1
101742    2
101743    1
101744    1
101745    2
101746    1
101747    1
101748    1
101749    3
101750    3
101751    3
101752    3
101753    1
101754    1
101755    1
101756    1
101757    1
101758    1
101759    1
101760    1
101761    1
101762    1
101763    1
101764    2
101765    1
Name: admission_type_id, Length: 101766, dtype: int64

In [21]:
df_2 = df.join(admission_type_id_df.set_index("admission_type_id"), on=["admission_type_id"])
df_2

ValueError: You are trying to merge on int64 and object columns. If you wish to proceed you should use pd.concat

### Step 2. Remove missing/unclear values.


In [ ]:
df = df[df['gender']!='Unknown/Invalid']
df = df[df['race']!='?']
df = df[df['admission_type_id'] != '6']
df = df[df['admission_type_id'] != 'NaN']
df = df[df['discharge_disposition_id'] != '18']
df = df[df['discharge_disposition_id'] != 'NaN']
df = df.drop(['citoglipton'], axis=1) #citoglipton got dropped because it did not prescribed to anyone

### Step 3. Renaming columns

In [ ]:
df.columns

## Data Cleaning and Transformation conclusion
At first glance, we can notice that the majority of people being re-admitted are:
- Caucasian
- Female
- Aged 50+
- takes between 9 and 17 medications
# TO BE FINISHED

In [22]:
df.to_csv("cleaned_data_v4.csv")